File: src/java/com/churchapp/service/AuthenticationService.java
package com.churchapp.service;

import com.churchapp.dto.auth.AuthenticationResponse;
import com.churchapp.dto.auth.LoginRequest;
import com.churchapp.dto.auth.RegisterRequest;
import com.churchapp.entity.User;
import com.churchapp.repository.MemberRepository;
import com.churchapp.repository.UserRepository;
import com.churchapp.security.ChurchUserPrincipal;
import com.churchapp.security.JwtService;
import io.vavr.control.Option;
import io.vavr.control.Try;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class AuthenticationService {

    private final UserRepository userRepository;
    private final MemberRepository memberRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;

    public Try<AuthenticationResponse> register(RegisterRequest request) {
        return Try.of(() -> {
            // Check if username already exists
            if (userRepository.existsByUsername(request.getUsername())) {
                throw new RuntimeException("Username already exists");
            }
            
            // Check if email already exists
            if (userRepository.existsByEmail(request.getEmail())) {
                throw new RuntimeException("Email already exists");
            }

            // Build user entity
            User.UserBuilder userBuilder = User.builder()
                .username(request.getUsername())
                .password(passwordEncoder.encode(request.getPassword()))
                .email(request.getEmail())
                .role(request.getRole())
                .active(true);

            // Link to member if provided
            if (request.getMemberId() != null) {
                return memberRepository.findById(request.getMemberId())
                    .map(member -> {
                        User user = userBuilder.member(member).build();
                        User savedUser = userRepository.save(user);
                        return generateAuthResponse(savedUser);
                    })
                    .orElseThrow(() -> new RuntimeException("Member not found"));
            } else {
                User user = userBuilder.build();
                User savedUser = userRepository.save(user);
                return generateAuthResponse(savedUser);
            }
        })
        .onFailure(throwable -> log.error("Failed to register user", throwable));
    }

    public Try<AuthenticationResponse> authenticate(LoginRequest request) {
        return Try.of(() -> {
            authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                    request.getUsername(),
                    request.getPassword()
                )
            );

            User user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() -> new RuntimeException("User not found"));

            return generateAuthResponse(user);
        })
        .onFailure(throwable -> log.error("Failed to authenticate user", throwable));
    }

    public Try<AuthenticationResponse> refreshToken(String refreshToken) {
        return Try.of(() -> {
            String username = jwtService.extractUsername(refreshToken);
            
            if (username == null) {
                throw new RuntimeException("Invalid refresh token");
            }

            User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

            ChurchUserPrincipal userPrincipal = new ChurchUserPrincipal(user);
            
            if (jwtService.isTokenValid(refreshToken, userPrincipal)) {
                return generateAuthResponse(user);
            } else {
                throw new RuntimeException("Invalid refresh token");
            }
        })
        .onFailure(throwable -> log.error("Failed to refresh token", throwable));
    }

    public Option<User> getCurrentUser(Authentication authentication) {
        return Try.of(() -> {
            if (authentication != null && authentication.getPrincipal() instanceof ChurchUserPrincipal) {
                ChurchUserPrincipal userPrincipal = (ChurchUserPrincipal) authentication.getPrincipal();
                return Option.of(userPrincipal.getUser());
            }
            return Option.<User>none();
        })
        .onFailure(throwable -> log.error("Failed to get current user", throwable))
        .getOrElse(Option.none());
    }

    private AuthenticationResponse generateAuthResponse(User user) {
        ChurchUserPrincipal userPrincipal = new ChurchUserPrincipal(user);
        String accessToken = jwtService.generateToken(userPrincipal);
        String refreshToken = jwtService.generateRefreshToken(userPrincipal);

        return AuthenticationResponse.builder()
            .accessToken(accessToken)
            .refreshToken(refreshToken)
            .tokenType("Bearer")
            .expiresIn(86400000L) // 24 hours in milliseconds
            .username(user.getUsername())
            .role(user.getRole().name())
            .build();
    }
}

File: src/kotlin/com/churchapp/service/AuthenticationService.kt
package com.churchapp.service

import arrow.core.Either
import arrow.core.None
import arrow.core.Option
import arrow.core.Some
import arrow.core.left
import arrow.core.right
import com.churchapp.entity.User
import com.churchapp.entity.enums.RoleType
import com.churchapp.repository.UserRepository
import com.churchapp.security.JwtTokenService
import org.slf4j.LoggerFactory
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class AuthenticationService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtTokenService: JwtTokenService,
) {
    private val logger = LoggerFactory.getLogger(AuthenticationService::class.java)

    // Authenticate user with functional error handling
    fun authenticate(
        username: String,
        password: String,
    ): Either<AuthError, AuthResult> {
        val credentialsResult = validateCredentials(username, password)
        return when (credentialsResult) {
            is Either.Left -> credentialsResult
            is Either.Right -> {
                val credentials = credentialsResult.value
                val userOption = findUserByUsername(credentials.username)
                when (userOption) {
                    is None -> AuthError.InvalidCredentials("User not found").left()
                    is Some -> {
                        val user = userOption.value
                        if (passwordEncoder.matches(credentials.password, user.password)) {
                            val tokenResult = generateToken(user)
                            when (tokenResult) {
                                is Either.Left -> tokenResult
                                is Either.Right -> AuthResult(user, tokenResult.value).right()
                            }
                        } else {
                            AuthError.InvalidCredentials("Invalid password").left()
                        }
                    }
                }
            }
        }
    }

    // Register new user with validation
    fun registerUser(
        username: String,
        email: String,
        password: String,
        role: RoleType = RoleType.MEMBER,
    ): Either<AuthError, User> {
        val validationResult = validateRegistration(username, email, password)
        return when (validationResult) {
            is Either.Left -> validationResult
            is Either.Right -> {
                val (_, _, _, normUsername, normEmail, normPassword) = validationResult.value

                if (userRepository.existsByUsername(normUsername)) {
                    AuthError.UserAlreadyExists("Username already exists").left()
                } else if (userRepository.existsByEmail(normEmail)) {
                    AuthError.UserAlreadyExists("Email already exists").left()
                } else {
                    try {
                        val encodedPassword = passwordEncoder.encode(normPassword)
                        val user =
                            User(
                                username = normUsername,
                                password = encodedPassword,
                                email = normEmail,
                                role = role,
                                createdAt = LocalDateTime.now(),
                            )
                        userRepository.save(user).right()
                    } catch (e: org.springframework.dao.DataIntegrityViolationException) {
                        // Handle race condition where username or email was taken after our check
                        logger.warn("Unique constraint violation registering user: $username", e)
                        AuthError.UserAlreadyExists("Username or email already exists").left()
                    } catch (e: Exception) {
                        logger.error("Error registering user: $username", e)
                        AuthError.RegistrationFailed(e.message ?: "Unknown error").left()
                    }
                }
            }
        }
    }

    // Generate JWT token
    fun generateToken(user: User): Either<AuthError, String> =
        jwtTokenService.generateToken(user)
            .mapLeft<AuthError> { AuthError.TokenGenerationFailed(it) }

    // Get user from token
    fun getUserFromToken(token: String): Either<AuthError, User> {
        val tokenValidationResult = jwtTokenService.validateToken(token)
        return when (tokenValidationResult) {
            is Either.Left -> AuthError.InvalidToken(tokenValidationResult.value).left()
            is Either.Right -> {
                val claims = tokenValidationResult.value
                val username = claims.subject
                val userOption = findUserByUsername(username)
                when (userOption) {
                    is None -> AuthError.UserNotFound("User not found: $username").left()
                    is Some -> userOption.value.right()
                }
            }
        }
    }

    // Find user by username with Option
    private fun findUserByUsername(username: String): Option<User> =
        try {
            Option.fromNullable(userRepository.findByUsername(username))
        } catch (e: Exception) {
            logger.error("Error finding user by username: $username", e)
            None
        }

    // Validation functions using Arrow
    private fun validateCredentials(
        username: String,
        password: String,
    ): Either<AuthError, Credentials> =
        when {
            username.isBlank() -> AuthError.ValidationError("Username cannot be blank").left()
            password.isBlank() -> AuthError.ValidationError("Password cannot be blank").left()
            else -> Credentials(username, password).right()
        }

    private fun validateRegistration(
        username: String,
        email: String,
        password: String,
    ): Either<AuthError, ValidationResult> {
        val normalizedUsername = username.trim()
        val normalizedEmail = email.trim()
        val normalizedPassword = password.trim()

        return when {
            normalizedUsername.isBlank() -> AuthError.ValidationError("Username cannot be blank").left()
            normalizedUsername.length < 3 -> AuthError.ValidationError("Username must be at least 3 characters").left()
            normalizedEmail.isBlank() -> AuthError.ValidationError("Email cannot be blank").left()
            !isValidEmail(normalizedEmail) -> AuthError.ValidationError("Invalid email format").left()
            normalizedPassword.isBlank() -> AuthError.ValidationError("Password cannot be blank").left()
            normalizedPassword.length < 6 -> AuthError.ValidationError("Password must be at least 6 characters").left()
            else ->
                ValidationResult(
                    username = normalizedUsername,
                    email = normalizedEmail,
                    password = normalizedPassword,
                ).right()
        }
    }

    private fun isValidEmail(email: String): Boolean = email.matches(Regex("^[A-Za-z0-9+_.-]+@([A-Za-z0-9.-]+\\.[A-Za-z]{2,})$"))
}

// Sealed class for authentication errors
sealed class AuthError(val message: String) {
    data class InvalidCredentials(val details: String) : AuthError("Invalid credentials: $details")

    data class UserAlreadyExists(val details: String) : AuthError("User already exists: $details")

    data class UserNotFound(val details: String) : AuthError("User not found: $details")

    data class ValidationError(val details: String) : AuthError("Validation error: $details")

    data class RegistrationFailed(val details: String) : AuthError("Registration failed: $details")

    data class TokenGenerationFailed(val details: String) : AuthError("Token generation failed: $details")

    data class InvalidToken(val details: String) : AuthError("Invalid token: $details")
}

// Data classes for authentication
data class Credentials(val username: String, val password: String)

data class ValidationResult(
    val username: String,
    val email: String,
    val password: String,
    val normalizedUsername: String = username.trim().lowercase(),
    val normalizedEmail: String = email.trim().lowercase(),
    val normalizedPassword: String = password.trim(),
)

data class AuthResult(val user: User, val token: String)
